src/main/java/org/traccar/protocol/LaipacProtocolDecoder.java

```diff
@@ -81,81 +100,166 @@ private String decodeAlarm(String event) {
+                int inputStatusInt = inputStatus - 'A';
+                position.set(Position.PREFIX_IN + 1, inputStatusInt & 1);
+                position.set(Position.PREFIX_IN + 2, inputStatusInt & 2);
+                position.set(Position.KEY_IGNITION, ((inputStatusInt & 1) != 0) ? true : false);
+                return null;
+            }
+        }
 
-        if (sentence.startsWith("$ECHK") && channel != null) {
-            channel.writeAndFlush(new NetworkMessage(sentence + "\r\n", remoteAddress)); // heartbeat
-            return null;
+        return event;
+    }
+
+    private String getDevicePassword(DeviceSession deviceSession) {
+
+        String devicePassword = defaultDevicePassword;
+
+        Device device = Context.getIdentityManager().getById(deviceSession.getDeviceId());
+        if (device != null) {
+            String password = device.getString(Command.KEY_DEVICE_PASSWORD);
+            if (password != null) {
+                devicePassword = password;
+            }
         }
 
-        Parser parser = new Parser(PATTERN, sentence);
-        if (!parser.matches()) {
-            return null;
+        return devicePassword;
+    }
+
+    private Object handleEchk(
+            String sentence, int checksum, Parser parser, Channel channel, SocketAddress remoteAddress) {
+
+        if (channel != null) {
+            channel.writeAndFlush(new NetworkMessage(sentence + "\r\n", remoteAddress));
         }
 
-        DeviceSession deviceSession = getDeviceSession(channel, remoteAddress, parser.next());
+        DeviceSession deviceSession =
+            getDeviceSession(channel, remoteAddress, parser.next());
+        if (deviceSession != null) {
+            ConnectionManager cm = Context.getConnectionManager();
+            if (cm != null) {
+                cm.updateDevice(deviceSession.getDeviceId(),
+                    Device.STATUS_ONLINE, new Date());
+            }
+        }
+
+        return null;
+    }
+
+    protected Object handleAvrmc(
+            String sentence, int checksum, Parser parser, Channel channel, SocketAddress remoteAddress) {
+
+        DeviceSession deviceSession =
+            getDeviceSession(channel, remoteAddress, parser.next());
         if (deviceSession == null) {
             return null;
         }
 
         Position position = new Position(getProtocolName());
+
+        // Device ID
         position.setDeviceId(deviceSession.getDeviceId());
 
+        // Time
         DateBuilder dateBuilder = new DateBuilder()
                 .setTime(parser.nextInt(0), parser.nextInt(0), parser.nextInt(0));
 
+        // Status [ V: Invalid | A: Valid | R: Not realtime | P: Parked ]
         String status = parser.next();
         String upperCaseStatus = status.toUpperCase();
-        position.setValid(upperCaseStatus.equals("A") || upperCaseStatus.equals("R") || upperCaseStatus.equals("P"));
+        position.setValid(upperCaseStatus.equals("A"));
```

Comment:
Why?

Reply:
V: Invalid | A: Valid | R: Not realtime | P: Parked
In tests I've done, I confirmed that the only reports that always are real (the device is in the position the reports says), are those marked with a 'a' or 'A'. You can take the device away to a real far place, removing the GPS antenna, and it will still send reports marked with an 'r' or 'R', using the same original position and creating false positives. What do you think about this? Thanks.

Reply:
Even if it's an old but valid GPS location, it should be marked as valid, assuming that timestamp is correct for the fix.

Reply:
I understand. Undoing this change. Thanks.